### PR TITLE
direct-to-display window improvements

### PR DESCRIPTION
**Pull request by gary-sweet**
## Description

Rather than just taking the first plane of the first display we now gather a list of valid candidates and pick the best fit.

An additional display 'FullscreenStretch' mode has been added. This uses VkDisplayPresentInfoKHR from the VK_KHR_display_swapchain extension, when available, to specify separate src and dst rectangles for the surface and display. This enables you to have a 640x360 surface stretched fullscreen onto the display for example, rather than just appearing in a corner.

In standard or stretch mode, with direct-to-display, specifying --width & --height controls the surface size.
In fullscreen or borderless modes, the surface is the same size as the display regardless of the width and height settings. The width and height are used to select the best display/plane combination however.

## General Checklist:

Please ensure the following points are checked:

- [x] My code follows the [coding style](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#Code-Style)
- [x] I have reviewed file [licenses](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#Copyright-Notice-and-License-Template)
- [x] I have updated the header Copyright to reflect the current year (CI build will fail if Copyright is out of date)
- [x] I have commented any added functions (in line with Doxygen)
- [x] I have commented any code that could be hard to understand
- [x] My changes do not add any new compiler warnings
- [x] My changes do not add any new validation layer errors or warnings
- [x] I have used existing framework/helper functions where possible
- [x] My changes build on Windows, Linux, macOS and Android. Otherwise I have [documented any exceptions](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#General-Requirements)
- [x] My changes do not add any regressions
- [x] I have tested every sample to ensure everything runs correctly
- [x] This PR describes the scope and expected impact of the changes I am making

Note: Built and tested only on Linux

### CODE REVIEW

File: framework/hpp_api_vulkan_sample.cpp [hunk 1]
```diff
@@ -485,6 +485,15 @@ void HPPApiVulkanSample::submit_frame()
 			present_info.setWaitSemaphores(semaphores.render_complete);
 		}
 
+		vk::DisplayPresentInfoKHR disp_present_info{};
```
Comment (asuessenbach) [inline framework/hpp_api_vulkan_sample.cpp]: You don't need to initialize the vk-structs with `{}`, as they have a default constructor.

File: framework/hpp_api_vulkan_sample.cpp [hunk 2]
```diff
@@ -485,6 +485,15 @@ void HPPApiVulkanSample::submit_frame()
...
+		if (device->is_extension_supported(VK_KHR_DISPLAY_SWAPCHAIN_EXTENSION_NAME) &&
+		    platform->get_window().get_display_present_info(reinterpret_cast<VkDisplayPresentInfoKHR *>(&disp_present_info),
+		                                                    extent.width, extent.height))
+		{
```
Comment (asuessenbach) [inline framework/hpp_api_vulkan_sample.cpp:491]: Would be great, if you could add some wrapper function to platform/hpp_window.h, like this:
```
	bool get_display_present_info(vk::DisplayPresentInfoKHR* info, uint32_t src_width, uint32_t src_height) const
	{
		return get_display_present_info(reinterpret_cast<VkDisplayPresentInfoKHR *>(info), src_width, src_height);
	}
```
to keep this file free from reinterpreting to VK-structs.

File: framework/platform/unix/direct_window.cpp [hunk 3]
```diff
@@ -244,76 +244,193 @@ VkSurfaceKHR DirectWindow::create_surface(Instance &instance)
...
+                                      std::max(properties.extent.width, best.caps.minSrcExtent.width));
+		image_extent.height = std::min(best.caps.maxSrcExtent.height,
+		                               std::max(properties.extent.height, best.caps.minSrcExtent.height));
 	}
 
-	VkExtent2D image_extent;
-	image_extent.width  = mode_props.parameters.visibleRegion.width;
-	image_extent.height = mode_props.parameters.visibleRegion.height;
+	// Calculate the display DPI
+	dpi = 25.4f * best.display_props.physicalResolution.width / best.display_props.physicalDimensions.width;
```
Comment (asuessenbach) [inline framework/platform/unix/direct_window.cpp]: Magic number `25.4f`? What is it?
Replace it with some define or const?

File: framework/hpp_api_vulkan_sample.cpp (hunk 1 above)
Comment (gary-sweet) [reply framework/hpp_api_vulkan_sample.cpp]: Done

File: framework/hpp_api_vulkan_sample.cpp (hunk 2 above)
Comment (gary-sweet) [reply framework/hpp_api_vulkan_sample.cpp:491]: Done

File: framework/platform/unix/direct_window.cpp (hunk 3 above)
Comment (gary-sweet) [reply framework/platform/unix/direct_window.cpp]: Done

File: framework/platform/hpp_window.h [hunk 4]
```diff
@@ -46,6 +46,11 @@ class HPPWindow : private vkb::Window
 	{
 		return static_cast<vk::SurfaceKHR>(create_surface(static_cast<VkInstance>(instance), static_cast<VkPhysicalDevice>(physical_device)));
 	}
+
+	bool get_display_present_info(vk::DisplayPresentInfoKHR *info, uint32_t src_width, uint32_t src_height) const
+	{
+		return vkb::Window::get_display_present_info(reinterpret_cast<VkDisplayPresentInfoKHR *>(info), src_width, src_height);
```
Comment (asuessenbach) [inline framework/platform/hpp_window.h]: Unfortunately, you should not explicitly call `vkb::Window::get_display_present_info`, as that would really just call that function.
Please add
`	using vkb::Window::get_display_present_info;`
to the using-section in this file, and remove the explicit `vkb::Window::` here to really get the actual override called.

Comment (gary-sweet) [reply framework/platform/hpp_window.h]: Thanks. That was a subtlety that had passed me by the first time around. Following what's actually going on with the class hierarchies is a lot trickier when they are reinterpret_cast'ed to other things.

**Comment by asuessenbach** [comment]
Besides the fact, that I have not checked the actual functionality (just checked on Windows where it is not used): LGTM now!

**Review by asuessenbach — APPROVED** [review]

**Comment by gary-sweet** [comment]
I suspect that no-one else will actually be able to test the direct-to-display changes, so this may be a code-review only change.

**Review by gpx1000 — APPROVED** [review]

File: framework/vulkan_sample.cpp [hunk 5]
```diff
@@ -131,6 +131,8 @@ bool VulkanSample::prepare(Platform &platform)
 
 	// Getting a valid vulkan surface from the platform
 	surface = platform.get_window().create_surface(*instance);
+	if (!surface)
+		throw std::runtime_error("Failed to create window surface.");
```
Comment (yangfengzzz) [inline framework/vulkan_sample.cpp:135]: Headless failed

Comment (gary-sweet) [reply framework/vulkan_sample.cpp:135]: This is a merged PR. If you believe there is a problem somewhere, please raise a new issue.

**Pull request merged**